lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -77,45 +96,210 @@ def __init__(
-        be completed to support additional combinations. I plan to support
-        additional flexibility in a separate PR based on issue #2370.
+        Generate the "query" node for an Elasticsearch document query.
+
+        In most cases this just adds a "term" within the "query"/"filter" node.
+        Several queries use an "and" term including user and access, which
+        generates two new "term" nodes. The most complicated is a query which
+        doesn't include user or access for an authorized user, which must
+        return all datasets "owned by the user or with access public".
+
+        Note that if a "user" JSON parameter is specified, the common
+        infrastructure has already checked that there's an authentication token
+        and that the token represents a user with access to that specified
+        user's data. This method is concerned only with constructing a proper
+        Elasticsearch query to represent the authenticated restrictions.
+
+        Specific cases for user and access values:
+
+            {}: defaulting both user and access
+                All private + public regardless of owner
+
+                ADMIN: all datasets
+                AUTHORIZED: all owner:mine OR access:public
+                UNAUTHORIZED: all access:public
+
+            {"user": "drb"}: defaulting access
+                All datasets owned by "drb"
+
+                ADMIN: all owner:drb
+                AUTHORIZED as drb: all owner:drb
+                UNAUTHORIZED (or non-drb): all owner:drb AND access:public
+
+            {"user": "drb, "access": "private"}: private drb
+                All datasets owned by "drb" with "private" access
+
+                ADMIN: all owner:drb AND access:private
+                AUTHORIZED as drb: all owner:drb AND access:private
+                UNAUTHORIZED (or non-drb): NO DATA
+
+            {"user": "drb", "access": "public"}: public drb
+                All datasets owned by "drb" with "public" access
+
+                ADMIN: all owner:drb AND access:public
+                AUTHORIZED as drb: all owner:drb AND access:public
+                UNAUTHORIZED (or non-drb): all owner:drb AND access:public
+
+                TODO: Need to decide if this is a security issue. Do we want to
+                hide the existence of other users here? E.g., should the
+                UNAUTHORIZED case return NO DATA? And if so, should there be a
+                difference between UNAUTHORIZED and AUTHORIZED for some other
+                user? (And, NOTE, specifying a "user" parameter, either without
+                authentication or with non-ADMIN authentication where the user
+                parameter doesn't match the specified "user" parameter, will
+                result in an UnverifiedUser exception during schema validation,
+                so we won't reach this point. That is, changes must be made if
+                we want to ALLOW these requests.)
```

Comment:
> Is a query for `wscales/private` a 400 or a 403?

I'm inclined to muddy the waters even further:

- if the query returns (or _would return_) hits, then it's neither a `400` nor a `403` (it's a `200` or similar), so we (should) only have to visit this issue for queries which return no hits;
- if the requestor is unauthorized, then the result should be same regardless of whether the user (or data) exists or not (but, I might lean toward a `401` rather than a `400` or a `403`);
- if the requestor is an Admin, then the result should be `404`, I would think (if the user or data doesn't exist);
- otherwise, I'm not sure whether the answer should be `403` (for security) or `404` ('cuz we trust the user).

Or, we could just punt:  we could always return `200`, and whether the response contains hits or not depends on whether the requestor had access -- a malicious attack will always see "success with no data", and there is no inference to be drawn (of course, this isn't particularly friendly to authenticated users, but it's safe and consistent).

> my inclination is to get `verify_user` out of the security business [...] leaving security concerns entirely to `ApiBase._check_authorization()`

I support that:  I'm a fan of separating mechanism from policy!

Reply:
The real question is whether we want to diagnose the client's right to ask the question. Sure, we can completely skip the CRUD model and just treat bad requests as no-op successes. Maybe that's not even entirely unreasonable for read-only operations like the Elasticsearch queries, although I'm not entirely happy with that.

We definitely don't want to distinguish between "that user doesn't exist" and "you can't ask about that user", so the 400 vs 403 discussion is important.

1. If I ask for Fred's public data, should I be able to distinguish between "Fred has no public data" and "there is no Fred"? (Maybe not.)
2. Same question if I ask for Fred's private data ... but I'm not entirely sure the answer is the same.

For example, maybe bad "public" data requests always return 200 with an empty payload, but bad "private" data requests always return 403? That is, one always has the *right* to ask for public data, though we don't want to explicitly reveal a bad username; but one does *not* have the right to ask for private data unless as admin or for your own data.

And 401 might be appropriate for an unauthenticated request that might have been OK if authenticated; e.g., if I ask for Dave's private data ... no-go unauthorized, but fine if authorized as Dave. It's not appropriate if authentication is given, but that authentication doesn't authorize the request (e.g., Dave asking for Webb's private data.) And I'm not sure I want to bother decoding 401 from 403 cases: I'm not sure it really has that much value in our context.

And I really am thinking about paring `verify_user` down to just basic validation of the `username` value, regardless of authorization. In fact, the `Auth` method should probably just be removed, and I could have `convert_username` simply do the query and either fail or return the `user.id`, leaving authorization entirely to `_check_authorization` where both the user and access can be considered. I'm going to do a commit of review comments first, but then I may try hacking around with this to see how much pain it'll cause.

Reply:
> If I ask for Fred's public data, should I be able to distinguish between "Fred has no public data" and "there is no Fred"? (Maybe not.)

If you ask for "Fred's public data" and the answer is "there is no public data for Fred", you cannot tell whether that's because there is no data or no Fred -- either way, the answer is correct...but, if we return something other than `200` or `404`, you'll be able to draw some sort of inference.  I don't think we want unauthorized users to be able to draw that inference; it would be nice if admin users could; and, for other users, personally I would support being able to help the user detect things like typos in usernames....

> Same question if I ask for Fred's private data ... but I'm not entirely sure the answer is the same.

Once again, if you get the answer "there is no private data for Fred", then you cannot draw any inference about the existence of Fred, unless we return something other than `200` or `404` (although in this case I think returning `401` might be an appropriate alternative if the requestor is unauthorized).  But, since we're talking about accessing something which is nominally "private", returning `403` to an authorized user might be appropriate; obviously, an admin would get `404` for an existent user and perhaps the same status with a different message for a non-existent one.

> I'm not sure I want to bother decoding 401 from 403 cases

I'm hoping that it will be obvious from context which one should be returned; I concur that it's not worth a lot of effort otherwise.  In fact, I'm at peace with returning `200`/`404` for all cases....

> I really am thinking about paring verify_user down to just basic validation of the username value, regardless of authorization.

Go you!  💪

Reply:
> to make sure that an unauthenticated user can never probe usernames by looking for `400` vs `401`, I've strictly outlawed any use of the `"user"` parameter for an unauthenticated user, even for public data.

I support the "minimal attack surface" approach, but I don't want us to cut off our nose to spite our face.  And, I don't know how much effort we want to put in on supporting non-logged-in users, but -- if we care -- a non-logged-in user is almost certainly going to want to search or sort or otherwise select from public data based on owner; so, if we don't provide that selection on the Server side, then we're going to force the Dashboard to fetch _all_ the public records and then filter the presentation to the end user in the browser.  That seems like a waste when we could easily do that filtering in the ES query and save a whole lot of data shuffling.

> On the whole, I don't like the idea of returning fake empty responses for disallowed queries. Again, if you're an unauthenticated client you're not even allowed to ask the question, regardless of whether the username exists or not. It's simple and consistent.

Returning "fake" empty responses is also simple and consistent; and, they aren't necessarily "fake" -- they are equivalent to what would have been returned if we had actually issued the ES query, we just took a shortcut.  My only (weak) concern about refusing to let the requestor ask the question is that I think it increases the data available from which to draw inferences...although, I believe I concur that it doesn't do so sufficiently to succeed.